internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala

```diff
@@ -183,9 +183,28 @@ object ClassToAPI {
     val fields = mergeMap(c, c.getDeclaredFields, c.getFields, fieldToDef(c, cf, enclPkg))
     val constructors =
       mergeMap(c, c.getDeclaredConstructors, c.getConstructors, constructorToDef(enclPkg))
+    val cl = c.getClassLoader
+    val name = c.getCanonicalName
+    val innerClasses = new mutable.ArrayBuffer[Class[_]]()
+    val publicClasses = new mutable.ArrayBuffer[Class[_]]()
+    val it = cf.innerClasses.toIterator
+    while (it.hasNext) {
+      val info = it.next
+      try {
+        if (info.outerClassName == name) {
+          val innerClass = cl.loadClass(info.innerClassName)
+          innerClasses += innerClass
+          if (info.isPublic) {
+            publicClasses += innerClass
+          }
+        }
+      } catch {
+        case _: ClassNotFoundException => // do nothing
+      }
+    }
     val classes = merge[Class[_]](c,
-                                  c.getDeclaredClasses,
-                                  c.getClasses,
+                                  innerClasses.toArray[Class[_]],
```

Comment:
@jvican 

> In all the examples I've seen, this weird bug seems to happen only in getClasses. Why are we re-implementing getDeclaredClasses too?

https://docs.oracle.com/javase/10/docs/api/java/lang/Class.html
> `getDeclaredClasses()` | Returns an array of `Class` objects reflecting all the classes and interfaces declared as members of the class represented by this Class object.
>
> `getClasses()` | Returns an array containing `Class` objects representing all the public classes and interfaces that are members of the class represented by this Class object.

These are more or less the same thing except one is public. I figured I would do a double duty here to save a reflection call.

Reply:
Is your implementation also reflecting all interface classes or only inner classes? e.g. maybe you can add a test proving that interfaces defined inside classes are detected by the new implementation.

Reply:
It's specified in the JVM spec, so I think we are good here. https://docs.oracle.com/javase/specs/jvms/se10/html/jvms-4.html#jvms-4.7.6

> Every CONSTANT_Class_info entry in the constant_pool table which represents a class **or interface** C that is not a package member must have exactly one corresponding entry in the classes array.

**Bold** added by me.